src/thunderbots/software/test/ai/hl/stp/stp.cpp

```diff
@@ -0,0 +1,161 @@
+    void SetUp() override
+    {
+        // Give an explicit seed to STP so that our tests are deterministic
+        stp   = STP(0);
+        world = ::Test::TestUtil::createBlankTestingWorld();
+    }
+
+    STP stp;
+    World world;
+};
+
+TEST_F(STPTest, test_only_test_plays_are_registered_in_play_factory)
+{
+    auto play_names = PlayFactory::getRegisteredPlayNames();
+    EXPECT_EQ(play_names.size(), 2);
+    EXPECT_EQ(std::count(play_names.begin(), play_names.end(), "Move Test Play"), 1);
+    EXPECT_EQ(std::count(play_names.begin(), play_names.end(), "Stop Test Play"), 1);
+}
+
+TEST_F(STPTest, test_exception_thrown_when_no_play_applicable)
+{
+    // Put the ball where both its x and y coordinates are negative. Neither test Play
+    // is applicable in this case
+    world = ::Test::TestUtil::setBallPosition(world, Point(-1, -1),
+                                              Timestamp::fromSeconds(0));
+    EXPECT_THROW(stp.calculateNewPlay(world), std::runtime_error);
+}
+
+TEST_F(STPTest, test_calculate_new_play_when_one_play_valid)
+{
+    // Only the StopTestPlay should be applicable
+    world =
+        ::Test::TestUtil::setBallPosition(world, Point(-1, 1), Timestamp::fromSeconds(0));
+    auto play = stp.calculateNewPlay(world);
+    EXPECT_TRUE(play);
+    EXPECT_EQ(play->name(), "Stop Test Play");
+}
+
+TEST_F(STPTest, test_calculate_new_play_when_multiple_plays_valid)
+{
+    // Both StopTestPlay and MoveTestPlay should be applicable
+    world =
+        ::Test::TestUtil::setBallPosition(world, Point(1, 1), Timestamp::fromSeconds(0));
+
+    // We expect a random selection of the plays that are applicable. This test should be
+    // deterministic because we have provided a seed for this test (in the setUp function)
+
+    std::unique_ptr<Play> play;
+    for (int i = 0; i < 8; i++)
+    {
+        play = stp.calculateNewPlay(world);
+        EXPECT_TRUE(play);
+        EXPECT_EQ(play->name(), "Move Test Play");
+    }
+
+    play = stp.calculateNewPlay(world);
+    EXPECT_TRUE(play);
+    EXPECT_EQ(play->name(), "Stop Test Play");
+
+    for (int i = 0; i < 2; i++)
+    {
+        play = stp.calculateNewPlay(world);
+        EXPECT_TRUE(play);
+        EXPECT_EQ(play->name(), "Move Test Play");
+    }
+
+    for (int i = 0; i < 3; i++)
+    {
+        play = stp.calculateNewPlay(world);
+        EXPECT_TRUE(play);
+        EXPECT_EQ(play->name(), "Stop Test Play");
+    }
+
+    play = stp.calculateNewPlay(world);
+    EXPECT_TRUE(play);
+    EXPECT_EQ(play->name(), "Move Test Play");
+
+    play = stp.calculateNewPlay(world);
+    EXPECT_TRUE(play);
+    EXPECT_EQ(play->name(), "Stop Test Play");
+
+    play = stp.calculateNewPlay(world);
+    EXPECT_TRUE(play);
+    EXPECT_EQ(play->name(), "Move Test Play");
+}
+
+TEST_F(STPTest, test_current_play_initially_unassigned)
+{
+    EXPECT_EQ(stp.getCurrentPlayName(), std::nullopt);
+}
+
+TEST_F(STPTest, test_play_assignment_transition_from_unassigned_to_assigned)
+{
```

Comment:
Quick comment on exactly what we're testing here would be good.

Reply:
comment added